File: src/CodexClient.cs
using System.ComponentModel;
using System.Diagnostics;
using System.Text.Json;

namespace CodexUsageMac;

internal sealed class CodexClient
{
    internal const long FiveHourMinutes = 300;
    internal const long WeeklyMinutes = 10_080;
    private static readonly TimeSpan RequestTimeout = TimeSpan.FromSeconds(15);

    internal async Task<UsageSnapshot> FetchAsync(CancellationToken cancellationToken = default)
    {
        if (!OperatingSystem.IsMacOS())
            throw new PlatformNotSupportedException("codex-usage-mac supports macOS only.");

        using CancellationTokenSource timeout = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
        timeout.CancelAfter(RequestTimeout);

        ProcessStartInfo startInfo = CreateStartInfo();
        using Process process = new() { StartInfo = startInfo };
        Task<string>? stderrTask = null;

        try
        {
            if (!process.Start())
                throw new InvalidOperationException("Could not start Codex CLI.");

            stderrTask = process.StandardError.ReadToEndAsync(timeout.Token);

            await SendAsync(process.StandardInput,
                """
                {"id":1,"method":"initialize","params":{"clientInfo":{"name":"codex-usage-mac","version":"1.0.0"},"capabilities":{"experimentalApi":true}}}
                """,
                timeout.Token).ConfigureAwait(false);

            string initialize = await WaitForResponseAsync(process.StandardOutput, 1, timeout.Token).ConfigureAwait(false);
            ThrowIfProtocolError(initialize);

            await SendAsync(process.StandardInput, """{"method":"initialized"}""", timeout.Token).ConfigureAwait(false);
            await SendAsync(process.StandardInput,
                """{"id":2,"method":"account/rateLimits/read","params":null}""",
                timeout.Token).ConfigureAwait(false);

            string response = await WaitForResponseAsync(process.StandardOutput, 2, timeout.Token).ConfigureAwait(false);
            ThrowIfProtocolError(response);
            return ParseUsageResponse(response, DateTimeOffset.Now);
        }
        catch (OperationCanceledException) when (!cancellationToken.IsCancellationRequested)
        {
            throw new InvalidOperationException(
                "Timed out while asking Codex for rate limits. Run `codex-usage doctor` and verify `codex login status`.");
        }
        catch (Win32Exception exception)
        {
            throw new InvalidOperationException(
                "Codex CLI was not found. Run `codex-usage doctor` for guided setup.", exception);
        }
        finally
        {
            try
            {
                process.StandardInput.Close();
            }
            catch
            {
            }

            if (!process.HasExited)
            {
                try { process.Kill(entireProcessTree: true); }
                catch { }
            }

            if (stderrTask is not null)
            {
                try { _ = await stderrTask.ConfigureAwait(false); }
                catch { }
            }
        }
    }

    private static ProcessStartInfo CreateStartInfo()
    {
        string executable = Environment.GetEnvironmentVariable("CODEX_CLI")?.Trim() ?? "codex";
        ProcessStartInfo info = new(executable)
        {
            CreateNoWindow = true,
            RedirectStandardError = true,
            RedirectStandardInput = true,
            RedirectStandardOutput = true,
            UseShellExecute = false,
        };
        info.ArgumentList.Add("app-server");
        info.ArgumentList.Add("--listen");
        info.ArgumentList.Add("stdio://");
        return info;
    }

    private static async Task SendAsync(StreamWriter writer, string message, CancellationToken cancellationToken)
    {
        await writer.WriteLineAsync(message.AsMemory(), cancellationToken).ConfigureAwait(false);
        await writer.FlushAsync(cancellationToken).ConfigureAwait(false);
    }

    private static async Task<string> WaitForResponseAsync(TextReader reader, int responseId, CancellationToken cancellationToken)
    {
        while (true)
        {
            string? line = await reader.ReadLineAsync(cancellationToken).ConfigureAwait(false);
            if (line is null)
                throw new InvalidOperationException("Codex exited before returning usage data.");

            JsonDocument? document = null;
            try
            {
                document = JsonDocument.Parse(line);
            }
            catch (JsonException)
            {
                continue;
            }

            using (document)
            {
                if (document.RootElement.TryGetProperty("id", out JsonElement idElement)
                    && idElement.ValueKind == JsonValueKind.Number
                    && idElement.TryGetInt32(out int id)
                    && id == responseId)
                {
                    return line;
                }
            }
        }
    }

    private static void ThrowIfProtocolError(string response)
    {
        using JsonDocument document = JsonDocument.Parse(response);
        if (!document.RootElement.TryGetProperty("error", out JsonElement error)) return;

        string message = error.TryGetProperty("message", out JsonElement detail)
            ? detail.GetString() ?? error.ToString()
            : error.ToString();
        throw new InvalidOperationException($"Codex app-server error: {message}");
    }

    internal static UsageSnapshot ParseUsageResponse(string response, DateTimeOffset refreshedAt)
    {
        using JsonDocument document = JsonDocument.Parse(response);
        JsonElement result = document.RootElement.GetProperty("result");
        JsonElement bucket = SelectCodexBucket(result);
        Dictionary<long, LimitWindow> windows = [];

        AddWindow(bucket, "primary", windows);
        AddWindow(bucket, "secondary", windows);

        windows.TryGetValue(FiveHourMinutes, out LimitWindow? fiveHour);
        windows.TryGetValue(WeeklyMinutes, out LimitWindow? weekly);
        return new UsageSnapshot(fiveHour, weekly, refreshedAt);
    }

    private static JsonElement SelectCodexBucket(JsonElement result)
    {
        if (result.TryGetProperty("rateLimitsByLimitId", out JsonElement buckets)
            && buckets.ValueKind == JsonValueKind.Object
            && buckets.TryGetProperty("codex", out JsonElement codex)
            && codex.ValueKind == JsonValueKind.Object)
        {
            return codex;
        }

        if (result.TryGetProperty("rateLimits", out JsonElement legacy)
            && legacy.ValueKind == JsonValueKind.Object)
        {
            return legacy;
        }

        throw new InvalidOperationException("Codex usage bucket was not present in the app-server response.");
    }

    private static void AddWindow(JsonElement bucket, string propertyName, IDictionary<long, LimitWindow> windows)
    {
        if (!bucket.TryGetProperty(propertyName, out JsonElement window)
            || window.ValueKind != JsonValueKind.Object
            || !window.TryGetProperty("windowDurationMins", out JsonElement durationElement)
            || !durationElement.TryGetInt64(out long duration)
            || !window.TryGetProperty("usedPercent", out JsonElement usedElement)
            || usedElement.ValueKind != JsonValueKind.Number)
        {
            return;
        }

        double used = Math.Clamp(usedElement.GetDouble(), 0d, 100d);
        windows[duration] = new LimitWindow(duration, used, ReadResetTime(window));
    }

    private static DateTimeOffset? ReadResetTime(JsonElement window)
    {
        if (!window.TryGetProperty("resetsAt", out JsonElement resetElement)
            || !resetElement.TryGetInt64(out long seconds))
        {
            return null;
        }

        try { return DateTimeOffset.FromUnixTimeSeconds(seconds).ToLocalTime(); }
        catch (ArgumentOutOfRangeException) { return null; }
    }
}

File: src/codex-client.mjs
import { spawn, spawnSync } from "node:child_process";
import { createInterface } from "node:readline";
import path from "node:path";
import process from "node:process";
import { FIVE_HOUR_MINUTES, WEEKLY_MINUTES } from "./usage.mjs";
import { VERSION } from "./meta.mjs";
import {
  desktopRuntimeCandidates,
  fileIsRunnable,
  findOnPath,
  makeCommandInvocation,
} from "./platform.mjs";

export function resolveCodex() {
  for (const variable of ["CODEX_CLI", "CODEX_CLI_PATH"]) {
    const override = process.env[variable]?.trim();
    if (!override) continue;
    return {
      command: findOnPath(override) ?? override,
      source: variable,
    };
  }

  const onPath = findOnPath("codex");
  if (onPath) return { command: onPath, source: "PATH" };

  for (const candidate of desktopRuntimeCandidates()) {
    if (fileIsRunnable(candidate.path)) {
      return { command: candidate.path, source: candidate.source };
    }
  }

  return {
    command: process.platform === "win32" ? "codex.exe" : "codex",
    source: "not found",
  };
}

export function resolveCodexCommand() {
  return resolveCodex().command;
}

export function codexExists() {
  const { command } = resolveCodex();
  if (path.isAbsolute(command) || command.includes("/") || command.includes("\\")) {
    return fileIsRunnable(command);
  }
  return Boolean(findOnPath(command));
}

export function codexVersion() {
  if (!codexExists()) return null;
  const invocation = makeCommandInvocation(resolveCodexCommand(), ["--version"]);
  const result = spawnSync(invocation.command, invocation.args, {
    encoding: "utf8",
    timeout: 5000,
    windowsHide: true,
  });
  return result.status === 0 ? (result.stdout || result.stderr).trim() : null;
}

export async function fetchUsage({ timeoutMs = 15000 } = {}) {
  const command = resolveCodexCommand();
  const invocation = makeCommandInvocation(command, ["app-server", "--stdio"]);
  const child = spawn(invocation.command, invocation.args, {
    stdio: ["pipe", "pipe", "pipe"],
    env: process.env,
    windowsHide: true,
  });

  let stderr = "";
  child.stderr.setEncoding("utf8");
  child.stderr.on("data", chunk => stderr += chunk);

  const rl = createInterface({ input: child.stdout, crlfDelay: Infinity });
  const waiting = new Map();
  let prematureExit = null;

  const finishWaiting = (err) => {
    for (const { reject, timer } of waiting.values()) {
      clearTimeout(timer);
      reject(err);
    }
    waiting.clear();
  };

  rl.on("line", line => {
    let message;
    try { message = JSON.parse(line); } catch { return; }
    if (message?.id != null && waiting.has(message.id)) {
      const item = waiting.get(message.id);
      clearTimeout(item.timer);
      waiting.delete(message.id);
      item.resolve(message);
    }
  });

  child.on("error", err => {
    prematureExit = err;
    finishWaiting(err);
  });
  child.on("exit", (code) => {
    if (waiting.size) {
      const err = new Error(`Codex exited before returning usage (exit ${code ?? "?"}). ${stderr.trim()}`.trim());
      prematureExit = err;
      finishWaiting(err);
    }
  });

  const request = (payload) => new Promise((resolve, reject) => {
    if (prematureExit) return reject(prematureExit);
    const timer = setTimeout(() => {
      waiting.delete(payload.id);
      reject(new Error("Timed out while waiting for Codex app-server."));
    }, timeoutMs);
    waiting.set(payload.id, { resolve, reject, timer });
    child.stdin.write(JSON.stringify(payload) + "\n");
  });

  try {
    const init = await request({
      id: 1,
      method: "initialize",
      params: { clientInfo: { name: "codex-usage", version: VERSION }, capabilities: { experimentalApi: true } },
    });
    throwProtocol(init);
    child.stdin.write(JSON.stringify({ method: "initialized" }) + "\n");
    const response = await request({ id: 2, method: "account/rateLimits/read", params: null });
    throwProtocol(response);
    return parseUsageResponse(response, new Date());
  } catch (err) {
    if (err?.code === "ENOENT") {
      const e = new Error("Codex runtime was not found. Install Codex CLI or a Codex desktop runtime.");
      e.code = "CODEX_NOT_FOUND";
      throw e;
    }
    throw err;
  } finally {
    rl.close();
    if (!child.killed) child.kill();
  }
}

function throwProtocol(response) {
  if (response?.error) throw new Error(`Codex API error: ${response.error.message ?? JSON.stringify(response.error)}`);
}

export function parseUsageResponse(response, refreshedAt = new Date()) {
  const result = response?.result;
  if (!result || typeof result !== "object") throw new Error("Codex returned an invalid rate-limit response.");
  const bucket = result.rateLimitsByLimitId?.codex ?? result.rateLimits;
  if (!bucket || typeof bucket !== "object") throw new Error("Codex general usage bucket was not found.");

  const windows = new Map();
  for (const name of ["primary", "secondary"]) {
    const raw = bucket[name];
    if (!raw || typeof raw.windowDurationMins !== "number" || typeof raw.usedPercent !== "number") continue;
    const reset = Number.isFinite(raw.resetsAt) ? new Date(raw.resetsAt * 1000) : null;
    windows.set(raw.windowDurationMins, {
      durationMinutes: raw.windowDurationMins,
      usedPercent: Math.max(0, Math.min(100, raw.usedPercent)),
      resetsAt: reset,
    });
  }
  return {
    fiveHour: windows.get(FIVE_HOUR_MINUTES) ?? null,
    weekly: windows.get(WEEKLY_MINUTES) ?? null,
    refreshedAt,
  };
}
